servicetalk-concurrent-api/src/test/java/io/servicetalk/concurrent/api/PublisherSwitchMapTest.java

```diff
@@ -0,0 +1,429 @@
+                });
+
+        Function<Integer, Publisher<? extends SwitchMapSignal<String>>> func = toSwitchFunction(i -> publisher2);
+        toSource(delayError ? publisher.switchMapDelayError(func, 2) : publisher.switchMap(func))
+                .subscribe(subscriber);
+
+        subscriber.awaitSubscription().request(1);
+        testSubscription.awaitRequestN(1);
+        publisher.onNext(0);
+
+        testSubscription2.awaitRequestN(1);
+        publisher2.onNext("foo");
+        validateSignal(subscriber.takeOnNext(), "foo", false);
+
+        if (delayError) {
+            DeliberateException deliberateException = new DeliberateException();
+            if (rootErrorFirst) {
+                publisher.onError(deliberateException);
+            }
+
+            Throwable secondCause = new IllegalStateException("second exception");
+            if (onComplete) {
+                publisher2.onComplete();
+            } else {
+                publisher2.onError(secondCause);
+            }
+
+            if (rootErrorFirst) {
+                Throwable cause = subscriber.awaitOnError();
+                assertThat(cause, is(deliberateException));
+                if (!onComplete) {
+                    assertThat(asList(cause.getSuppressed()), contains(secondCause));
+                }
+            } else {
+                publisher.onError(deliberateException);
+                Throwable cause = subscriber.awaitOnError();
+                if (onComplete) {
+                    assertThat(cause, is(deliberateException));
+                } else {
+                    assertThat(cause, is(secondCause));
+                    assertThat(asList(cause.getSuppressed()), contains(deliberateException));
+                }
+            }
+            assertThat(testSubscription2.isCancelled(), equalTo(false));
+        } else {
+            publisher.onError(DELIBERATE_EXCEPTION);
+            testSubscription2.awaitCancelled();
+            assertThat(subscriber.awaitOnError(), is(DELIBERATE_EXCEPTION));
+        }
+    }
+
+    private static <R> void validateSignal(@Nullable SwitchMapSignal<R> signal, R data, boolean isSwitched) {
+        assertThat(signal, notNullValue());
+        assertThat(signal.isSwitched(), equalTo(isSwitched));
+        assertThat(signal.data(), equalTo(data));
+    }
+
+    private static <P, S> void validateTerminal(TestPublisher<P> publisher, TestPublisherSubscriber<S> subscriber,
+                                                boolean onError) {
+        if (onError) {
+            publisher.onError(DELIBERATE_EXCEPTION);
+            assertThat(subscriber.awaitOnError(), is(DELIBERATE_EXCEPTION));
+        } else {
+            publisher.onComplete();
+            subscriber.awaitOnComplete();
+        }
+    }
+
+    /**
+     * A signal containing the data from a series of {@link Publisher}s connected in a serial fashion.
+     * @param <T> Type of the data.
+     */
+    interface SwitchMapSignal<T> {
```

Comment:
This has shown to be useful but will keep internal to test for now.

When a switch happens application logic may need to take special action (e.g. replay initial/aggregated state as first event)